### PR TITLE
add ghcr

### PR DESCRIPTION
**Pull request by ksurl**
cleaned up version of https://github.com/containrrr/watchtower/pull/808

PR for https://github.com/containrrr/watchtower/issues/722

## PREREQUISITES  
* create github personal access token with permissions for read/write packages as seen [here](https://docs.github.com/en/packages/guides/migrating-to-github-container-registry-for-docker-images)
* add the token as a repo environment variable named GITHUB_PASSWORD
* enable enhanced container support in profile settings (optional?)

## POST MERGE  
* after first build has been published, package needs to be made public (defaults to private)

### CODE REVIEW

**Comment by piksel** [comment]
Haha, the name of the secret needs to be changed. Github did not like that one :D
image

File: .github/workflows/release-dev.yaml [hunk 1]
```diff
@@ -21,3 +21,12 @@ jobs:
         file: dockerfiles/Dockerfile.self-contained
         repository: containrrr/watchtower
         tags: latest-dev
+
+    - uses: jerray/publish-docker-action@master
+      with:
+        username: ${{ github.repository_owner }}
+        password: ${{ secrets.GITHUB_PASSWORD }}
```
Comment (piksel) [inline .github/workflows/release-dev.yaml]:
```suggestion
        username: ${{ secrets.BOT_USERNAME }}
        password: ${{ secrets.BOT_GHRC_PAT }}
```

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -141,6 +163,11 @@ jobs:
           echo "$DOCKER_PASS" | docker login -u $DOCKER_USER --password-stdin && \
             docker manifest push containrrr/watchtower:$(echo $TAG | sed 's/^v*//') && \
             docker manifest push containrrr/watchtower:latest
+      - name: Push manifests to GitHub Container Registry
+        run: |
+          echo "$GITHUB_PASSWORD" | docker login -u ${{ github.repository_owner }} --password-stdin && \
```
Comment (piksel) [inline .github/workflows/release.yml]:
```suggestion
          echo "$BOT_GHRC_PAT" | docker login -u $BOT_USERNAME --password-stdin && \
```

**Review by piksel — CHANGES_REQUESTED** [review]
Suggested the configured secrets

File: .github/workflows/release-dev.yaml (hunk 1 above)
Comment (simonaronsson) [reply .github/workflows/release-dev.yaml]: @piksel `BOT_GHRC_PAT` -> `BOT_GHCR_PAT` though :)

File: .github/workflows/release.yml (hunk 2 above)
Comment (simonaronsson) [reply .github/workflows/release.yml]: same here

**Comment by piksel** [comment]
Tested just pushing the current `latest` to GHRC using the credentials above. It works, and should work with this action as well.

**Comment by ksurl** [comment]
looks good. once it is merged, it should start pushing all new versioned and latest builds to it and mirror docker hub?

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -141,6 +163,11 @@ jobs:
           echo "$DOCKER_PASS" | docker login -u $DOCKER_USER --password-stdin && \
             docker manifest push containrrr/watchtower:$(echo $TAG | sed 's/^v*//') && \
             docker manifest push containrrr/watchtower:latest
+      - name: Push manifests to GitHub Container Registry
+        run: |
+          echo "$BOT_GHRC_PAT" | docker login -u $BOT_USERNAME --password-stdin && \
```
Comment (piksel) [inline .github/workflows/release.yml]:
```suggestion
          echo "$BOT_GHCR_PAT" | docker login -u $BOT_USERNAME --password-stdin && \
```

File: .github/workflows/release-dev.yaml [hunk 4]
```diff
@@ -21,3 +21,12 @@ jobs:
         file: dockerfiles/Dockerfile.self-contained
         repository: containrrr/watchtower
         tags: latest-dev
+
+    - uses: jerray/publish-docker-action@master
+      with:
+        username: ${{ secrets.BOT_USERNAME }}
+        password: ${{ secrets.BOT_GHRC_PAT }}
```
Comment (piksel) [inline .github/workflows/release-dev.yaml]:
```suggestion
        password: ${{ secrets.BOT_GHCR_PAT }}
```

**Review by piksel — APPROVED** [review]

File: .github/workflows/release.yml [hunk 5]
```diff
@@ -123,24 +134,35 @@ jobs:
...
             --arch arm
     
           docker manifest annotate \
             containrrr/watchtower:latest \
             containrrr/watchtower:arm64v8-latest \
+            ghcr.io/containrrr/watchtower:latest \
+            ghcr.io/containrrr/watchtower:arm64v8-latest \
             --os linux \
             --arch arm64 \
             --variant v8
```
Comment (piksel) [inline .github/workflows/release.yml]:
```suggestion
      - name: Annotate manifest for latest
        run: |
          for REPO in '' ghrc.io/ ; do

          docker manifest annotate \
            ${REPO}containrrr/watchtower:latest \
            ${REPO}containrrr/watchtower:i386-latest \
            --os linux \
            --arch 386
      
          docker manifest annotate \
            ${REPO}containrrr/watchtower:latest \
            ${REPO}containrrr/watchtower:armhf-latest \
            --os linux \
            --arch arm
            
          docker manifest annotate \
            ${REPO}containrrr/watchtower:latest \
            ${REPO}containrrr/watchtower:arm64v8-latest \
            --os linux \
            --arch arm64 \
            --variant v8

          done
```

File: .github/workflows/release.yml [hunk 6]
```diff
@@ -95,24 +95,35 @@ jobs:
...
             --arch arm
       
           docker manifest annotate \
             containrrr/watchtower:$(echo $TAG | sed 's/^v*//') \
             containrrr/watchtower:arm64v8-$(echo $TAG | sed 's/^v*//') \
+            ghcr.io/containrrr/watchtower:$(echo $TAG | sed 's/^v*//') \
+            ghcr.io/containrrr/watchtower:arm64v8-$(echo $TAG | sed 's/^v*//') \
             --os linux \
             --arch arm64 \
             --variant v8
```
Comment (piksel) [inline .github/workflows/release.yml]:
```suggestion
        run: |
          for REPO in '' ghrc.io/ ; do
          
          docker manifest annotate \
            ${REPO}containrrr/watchtower:$(echo $TAG | sed 's/^v*//') \
            ${REPO}containrrr/watchtower:i386-$(echo $TAG | sed 's/^v*//') \
            --os linux \
            --arch 386
          
          docker manifest annotate \
            ${REPO}containrrr/watchtower:$(echo $TAG | sed 's/^v*//') \
            ${REPO}containrrr/watchtower:armhf-$(echo $TAG | sed 's/^v*//') \
            --os linux \
            --arch arm
      
          docker manifest annotate \
            ${REPO}containrrr/watchtower:$(echo $TAG | sed 's/^v*//') \
            ${REPO}containrrr/watchtower:arm64v8-$(echo $TAG | sed 's/^v*//') \
            --os linux \
            --arch arm64 \
            --variant v8
            
            done
```

**Review by piksel — CHANGES_REQUESTED** [review]
The annotations needs to be done for every repo and cannot be altered to include both at the same time. They link `tag_from_arg1` to `tag_from_arg2`, so providing four arguments doesn't do what was intended.

**Review by simonaronsson — APPROVED** [review]

**Comment by simonaronsson** [comment]
@all-contributors add @ksurl for code

**Pull request merged**